### PR TITLE
Laboratory work #1, Elena Ermak - 22FPL1

### PR DESCRIPTION
**Pull request by ermakelena**

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -11,6 +11,16 @@ def tokenize(text: str) -> list[str] | None:
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
+    tokens = []
+    if isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:15]: please, use early return

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -11,6 +11,16 @@ def tokenize(text: str) -> list[str] | None:
...
+        split_text = list(text.lower())
+
+        for i in split_text:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:18]: try to use list comprehension

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -19,6 +29,17 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
+    freq_dict = {}
+    if isinstance(tokens, list):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:33]: again - early return

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -29,6 +50,15 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
     :return: a dictionary with two keys – name, freq
     """
 
+    if isinstance(language, str) and isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:53]: again - early return

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Good start! 
Please, remove lw.py

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py:15]: done

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py:18]: done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py:33]: done

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py:53]: done

File: README.md [hunk 5]
```diff
@@ -1,95 +0,0 @@
-# Лабораторные работы для 2-го курса ФПЛ (2023/2024)
```
Comment (artyomtugaryov) [inline README.md:1]: Why did you remove it? Good useful file...

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -12,13 +12,28 @@ def tokenize(text: str) -> list[str] | None:
     :return: a list of lower-cased tokens without punctuation
     """
 
+    if not isinstance(text, str):
+        return None
+
+    split_text = list(text.lower())
+    tokens = [i for i in split_text if i.isalpha()]
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:19]: you can return without creating the variable

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -12,13 +12,28 @@ def tokenize(text: str) -> list[str] | None:
...
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
+    if not isinstance(tokens, list):
+        return None
+
+    for i in tokens:
+        if not isinstance(i, str):
+            return None
+    freq_dict = {i: tokens.count(i) / len(tokens) for i in tokens}
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:35]: again

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -29,6 +44,16 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
     :return: a dictionary with two keys – name, freq
     """
 
+    if not (isinstance(language, str) and isinstance(text, str)):
+        return None
+
+    tokens = tokenize(text)
+    profile = {
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:51]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -37,6 +62,15 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
     :return: the score
     """
+    if (
+            not (isinstance(predicted, list)
+                 and isinstance(actual, list))
+            or len(predicted) != len(actual)
+    ):
+        return None
+
+    mse = sum([(i - j) ** 2 for i, j in zip(actual, predicted)]) / len(actual)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:72]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -49,6 +83,19 @@ def compare_profiles(
...
+                 or isinstance(profile_to_compare, dict))
+            or ('name' or 'freq') not in unknown_profile
+            or ('name' or 'freq') not in profile_to_compare
+    ):
+        return None
+
+    all_tokens = set(unknown_profile['freq'].keys()) | set(profile_to_compare['freq'].keys())
+    language1 = [unknown_profile['freq'].get(token, 0) for token in all_tokens]
+    language2 = [profile_to_compare['freq'].get(token, 0) for token in all_tokens]
+    distance = calculate_mse(language1, language2)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:97]: and again

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -63,6 +110,24 @@ def detect_language(
...
+
+    if not (isinstance(distance1, float)
+            and isinstance(distance2, float)):
+        return None
+
+    if distance1 > distance2:
+        return str(profile_2['name'])
+    if distance1 < distance2:
+        return str(profile_1['name'])
+    if distance1 == distance2:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:129]: I think you do not need the file

File: lab_1_classify_by_unigrams/target_score.txt [hunk 12]
```diff
@@ -1 +1 @@
-0
+8
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/target_score.txt:1]: Le't try 10!

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Need merge. Please, ask your mentor on the next seminar about merge

File: README.md (hunk 5 above)
Comment (ermakelena) [reply README.md]: I removed it by accident, you helped me to fix it in class. The file is very useful, i agree

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py:19]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py:35]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py:51]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py:72]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py:97]: fixed

File: docs/images/pycharm_control_tests_scope.jpg
Comment (artyomtugaryov) [inline docs/images/pycharm_control_tests_scope.jpg:1]: Please, revert the file!

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -12,13 +12,26 @@ def tokenize(text: str) -> list[str] | None:
     :return: a list of lower-cased tokens without punctuation
     """
 
+    if not isinstance(text, str):
+        return None
+
+    split_text = list(text.lower())
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:18]: Why do you make list there?

File: lab_1_classify_by_unigrams/target_score.txt (hunk 12 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/target_score.txt:1]: :(

File: seminars/practice_2_string.py [hunk 14]
```diff
@@ -1,200 +1,229 @@
...
-# string_match('xxcaazz', 'xxbaaz') → 3
-# NOTE: «xx», «aa» and «az» are found at the same position in both strings
-# string_match('abc', 'abc') → 2
-# string_match('abc', 'axc') → 0
-# string_match('he', 'hello') → 1
-# string_match('h', 'hello') → 0
-# string_match('aabbccdd', 'abbbxxd') → 1
-# string_match('aaxxaaxx', 'iaxxai') → 3
-# string_match('iaxxai', 'aaxxaaxx') → 3
+# """
```
Comment (artyomtugaryov) [inline seminars/practice_2_string.py:1]: revert, please

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Good, but revert unnecessary changes

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py]: fixed it, it was unnecessary

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -12,13 +13,24 @@ def tokenize(text: str) -> list[str] | None:
...
 def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     """
     Calculates frequencies of given tokens
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
+    if not isinstance(tokens, list):
+        return None
+
+    for i in tokens:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:30]: two loops. Try to make it in one

File: lab_1_classify_by_unigrams/main.py [hunk 16]
```diff
@@ -49,6 +87,18 @@ def compare_profiles(
...
+    if (
+            not (isinstance(unknown_profile, dict)
+                 or isinstance(profile_to_compare, dict))
+            or ('name' or 'freq') not in unknown_profile
+            or ('name' or 'freq') not in profile_to_compare
+    ):
+        return None
+
+    all_tokens = set(unknown_profile['freq'].keys()) | set(profile_to_compare['freq'].keys())
+    language1 = [unknown_profile['freq'].get(token, 0) for token in all_tokens]
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:99]: Again - too many loops

File: lab_1_classify_by_unigrams/main.py [hunk 17]
```diff
@@ -63,6 +113,24 @@ def detect_language(
...
+    if (
+            isinstance(distance1, float)
+            and isinstance(distance2, float)
+    ):
+
+        if distance1 > distance2:
+            return str(profile_2['name'])
+        if distance1 < distance2:
+            return str(profile_1['name'])
+    return [profile_1['name'], profile_2['name']].sort()
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:133]: Are you sure? Typing says that you can return only one string

File: lab_1_classify_by_unigrams/main.py [hunk 18]
```diff
@@ -80,6 +158,28 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
...
+        return None
+
+    preprocessed_profile = {
+        'name': profile['name'],
+        'freq': {}
+    }
+
+    for token in profile['freq']:
+        if token.lower() in preprocessed_profile['freq']:
+            preprocessed_profile['freq'][token.lower()] \
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:176]: you call `lower` thrice - save it to a variable, also it is more readable if you save to variables  `profile['freq'][token]` and  `profile['n_words'][0]`

File: seminars/practice_2_string.py (hunk 14 above)
Comment (artyomtugaryov) [reply seminars/practice_2_string.py:1]: If you do not know how - ask in chat. Or just copy the content of the file from https://github.com/fipl-hse/2023-2-level-labs/blob/main/seminars/practice_2_string.py

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nice! Fix small comments and revert seminar listings at last to main version:)

File: lab_1_classify_by_unigrams/main.py (hunk 18 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 16 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py:31]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 17 above)
Comment (ermakelena) [reply lab_1_classify_by_unigrams/main.py]: fixed
